### PR TITLE
Add kwarg build_by_default

### PR DESCRIPTION
**Pull request by jpakkane**
Targets that have `build_on_all` set to true are built by `ninja all` or equivalent. Those that have it set to false get built if the target is installed or used. There are no behavior changes. This allows two new things:

 - building custom targets by default
 - not building build targets by default (usually rarely used debug tools and the like)

The former is the main thing, the latter we get for free.

The kwarg name is perhaps not the best, suggestions for a better one welcome.

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -471,6 +471,17 @@ def test_compiler_cpp_stds(self):
         self._test_stds_impl(testdir, cpp, 'cpp')
 
 
+    def test_build_on_all(self):
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E303] too many blank lines (2)

File: mesonbuild/backend/ninjabackend.py [hunk 2]
```diff
@@ -2160,21 +2160,23 @@ def generate_utils(self, outfile):
         elem.add_item('pool', 'console')
         elem.write(outfile)
 
+    def get_build_on_all_targets(self):
+        result = []
+        for t in self.build.get_targets().values():
+            if t.build_on_all or \
+                (hasattr(t, 'install') and t.install) or\
+                (hasattr(t, 'build_always') and t.build_always):
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E129] visually indented line with same indent as next logical line

**Comment by tp-m** [comment]
`build_always` or `always_build` ?

Does a plain `ninja` default to the *all* target or is an explicit `ninja all` required for these to be built then?

**Comment by jpakkane** [comment]
Plain `ninja` internally invokes `ninja all`.

**Comment by nirbheek** [comment]
`build_always` already exists and has a meaning:

```
if true this target is always considered out of date and is rebuilt every time, useful for things such as build timestamps or revision control tags
```

How about `build_default : true`? People may not know the significance of the word `all` here, but the meaning of `default` will be clearer.

I really like this, btw, and the patch looks good!

**Comment by jpakkane** [comment]
`build_default` is ambigious, it should really be `build_by_default`, but that is too long ...

**Pull request merged**